gk/fib.c

```diff
@@ -1664,3 +1666,300 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
 
 	return del_fib_entry_numerical(&prefix_info, gk_conf);
 }
+
+static void
+fillup_gk_fib_dump_entry_ether(struct gk_fib_dump_entry *dentry,
+	struct ether_cache *eth_cache)
+{
+	dentry->stale = eth_cache->stale;
+	rte_memcpy(&dentry->nexthop_ip, &eth_cache->ip_addr,
+		sizeof(dentry->nexthop_ip));
+	rte_memcpy(&dentry->d_addr, &eth_cache->l2_hdr.eth_hdr.d_addr,
+		sizeof(dentry->d_addr));
+}
+
+static void
+fillup_gk_fib_dump_entry(struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	dentry->action = fib->action;
+	switch (fib->action) {
+	case GK_FWD_GRANTOR:
+		rte_memcpy(&dentry->grantor_ip, &fib->u.grantor.gt_addr,
+			sizeof(dentry->grantor_ip));
+		fillup_gk_fib_dump_entry_ether(dentry,
+			fib->u.grantor.eth_cache);
+		break;
+
+	case GK_FWD_GATEWAY_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_GATEWAY_BACK_NET:
+		fillup_gk_fib_dump_entry_ether(dentry,
+			fib->u.gateway.eth_cache);
+		break;
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		/* FALLTHROUGH */
+	case GK_DROP:
+		break;
+
+	default:
+		rte_panic("Invalid FIB action (%u) at %s with lcore %u\n",
+			fib->action, __func__, rte_lcore_id());
+		break;
+	}
+}
+
+#define CTYPE_STRUCT_FIB_DUMP_ENTRY_PTR "struct gk_fib_dump_entry *"
+
+static void
+list_ipv4_fib_entries(lua_State *l, struct gk_lpm *ltbl)
+{
+	int ret, index;
+
+	static bool assigned_type_fib_dump_entry = false;
+	static uint32_t correct_ctypeid_fib_dump_entry;
+
+	struct gk_fib *fib;
+	const struct rte_lpm_rule *re4;
+	struct rte_lpm_iterator_state state;
+	struct gk_fib_dump_entry dentry;
+	void *cdata;
+
+	if (!assigned_type_fib_dump_entry) {
+		correct_ctypeid_fib_dump_entry = luaL_get_ctypeid(l,
+			CTYPE_STRUCT_FIB_DUMP_ENTRY_PTR);
+		assigned_type_fib_dump_entry = true;
+	}
+
+	ret = rte_lpm_iterator_state_init(ltbl->lpm, 0, 0, &state);
+	if (ret < 0)
+		luaL_error(l, "gk: failed to initialize the lpm rule iterator state at %s!",
+			__func__);
+
+	index = rte_lpm_rule_iterate(&state, &re4);
+	while (index >= 0) {
+		fib = &ltbl->fib_tbl[re4->next_hop];
+		if (fib->action == GK_FWD_NEIGHBOR_FRONT_NET ||
+				fib->action == GK_FWD_NEIGHBOR_BACK_NET) {
+			index = rte_lpm_rule_iterate(&state, &re4);
+			continue;
+		}
+
+		memset(&dentry, 0, sizeof(dentry));
+		dentry.addr.proto = ETHER_TYPE_IPv4;
+		dentry.addr.ip.v4.s_addr = htonl(re4->ip);
+		dentry.prefix_len = state.depth;
+		fillup_gk_fib_dump_entry(&dentry, fib);
+
+		lua_pushvalue(l, 2);
+		lua_insert(l, 3);
+		cdata = luaL_pushcdata(l, correct_ctypeid_fib_dump_entry,
+			sizeof(struct gk_fib_dump_entry *));
+		*(struct gk_fib_dump_entry **)cdata = &dentry;
+		lua_insert(l, 4);
+
+		if (lua_pcall(l, 2, 1, 0) != 0) {
```

Comment:
Since we are just raising the Lua error further here, we can drop the `if` statement and just rename `lua_pcall()` with `lua_call()`.